internal/http/server.go

```diff
@@ -24,15 +24,21 @@ func Serve(ctx context.Context, logger log.Logger, address string, handler http.
 		ReadHeaderTimeout: 20 * time.Second,
 	}
 
+	errChan := make(chan error, 1)
 	go func() {
 		logger.Info(fmt.Sprintf("Listening on %s", address))
 		if err := server.ListenAndServe(); err != nil && !errors.Is(err, http.ErrServerClosed) {
 			logger.Info(err.Error())
+			errChan <- err
 		}
 	}()
 
 	// Wait until we're told to shutdown.
-	<-ctx.Done()
+	select {
```

Comment:
This won't work, unfortunately, because of the pseudorandom case selection behavior of the `select` statement _and_ because we shutdown via `server.Shutdown` on l.50 so, at this point, nothing has told `ListenAndServe` to stop.

To retrieve the error we'll probably want to funnel all errors into `errChan` including from the context timeout. We'd then launch the `Shutdown` in a separate goroutine and retrieve the error from the errChan only.

```go
errChan := make(chan error, 2)

// Launch ListenAndServe

<-ctx.Done()

ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
defer cancel()

go func() {
	errChan <- server.Shutdown(ctx)
}()

// We'll eithe rreceive an error from the ListenAndServe routine or
// a context related error from the timeout.
// For the timeout case we want to forcefully close the server and
// tidy the error but for all other cases we just want to return that error.
//
// Note we don't need a select statement because we _know_ we're using a 
// timeout context and if we trust the standard library we're guarenteeing 
// errChan will get at least 1 value.
err := <- errChan
if err != nil {
  if errors.Is(err, context.DeadlineExceeded) {
    server.Close()
    return errors.New("timed out waiting for graceful shutdown")
  }

  return err
}

return nil

Reply:
Apologies, I'm not following how this won't work. Both manual testing and the unit test have shown success. And it's only pseudorandom if both channels receive values at the same time. In that case it seems fine for either path to be selected.

For the shutdown, this code is exactly the same before this PR. That seems to function correctly, no?

Reply:
I thought on this more and I think the changes are fine as is. I was worried about missing an error if ctx.Done() is done, but frankly its irrelevant because the context is done indicating we should be shutting down so who cares if we miss the error 🤦🏻 .